### PR TITLE
Implement partitiontest_linter

### PR DESCRIPTION
**Pull request by algojack**
## Summary

Adding linter for partitiontest (as well as tests for the linter)

## Test Plan

Ran tests from VCS and from terminal using manual way in `cmd/partitiontest_linter`: `go test -v`
Output:
```
% go test -v
=== RUN   TestAll
--- PASS: TestAll (0.81s)
PASS
ok      command-line-arguments  1.244s
```

As well as run the linter manually in `cmd/partitiontest_linter/`: `go run plugin/plugin.go -- ./...` from root of repo.
Expected output is a list of files and functions with exit error code 3:
```
/Users/jack/projects/algojack.go-algorand/cmd/partitiontest_linter/linter_test.go:25:1: TestAll function is missing partitiontest.PartitionTest(<T type parameter>)
exit status 3
```

And finally ran them using `golangci-lint run` from root of repo:
```
% golangci-lint run
...
logging/telemetryhook_test.go:172:1: lint: TestAsyncTelemetryHook_CloseDrop function is missing partitiontest.PartitionTest(<T type parameter>) (partitiontest)
func TestAsyncTelemetryHook_CloseDrop(t *testing.T) {
^
logging/telemetryhook_test.go:197:1: lint: TestAsyncTelemetryHook_QueueDepth function is missing partitiontest.PartitionTest(<T type parameter>) (partitiontest)
func TestAsyncTelemetryHook_QueueDepth(t *testing.T) {
^
daemon/algod/api/server/v1/handlers/handlers_test.go:31:1: lint: TestDecorateUnknownTransactionTypeError function is missing partitiontest.PartitionTest(<T type parameter>) (partitiontest)
func TestDecorateUnknownTransactionTypeError(t *testing.T) {
^
test/e2e-go/cli/goal/expect/goal_expect_test.go:26:1: lint: TestGoalWithExpect function is missing partitiontest.PartitionTest(<T type parameter>) (partitiontest)
func TestGoalWithExpect(t *testing.T) {
^
...
```

### CODE REVIEW

**Comment by algojack** [comment]
not sure how i got 75 commits, but i'm guessing it can be squash merged to avoid long history

File: cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go
Comment (algorandskiy) [inline cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go:49]: remove or enable

File: cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go
Comment (algorandskiy) [inline cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go:6]: github.com/algorand/go-algorand/test/partitiontest

File: cmd/partitiontest_linter/analyzer/analyzer_test.go
Comment (algorandskiy) [inline cmd/partitiontest_linter/analyzer/analyzer_test.go:1]: why not `analyzer`? it is a unit test for a sibling file. By doing this 1) is it the way 2) no analyzer import needed below

**Review by algorandskiy — CHANGES_REQUESTED** [review]
Some coding style remarks. Plus I do not think we need to commit compiled binary here

File: cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go
Comment (algojack) [reply cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go:6]: this fails for me:
```
Running tool: /usr/local/go/bin/go test -timeout 30s -coverprofile=/var/folders/m7/v0208hbj5p7643101fcjkpl00000gp/T/vscode-go78tJNa/go-code-cover github.com/algorand/go-algorand/cmd/partitiontest_linter/analyzer

analyzer_testdata_test.go:7:2: cannot find package "github.com/algorand/go-algorand/test/partitiontest" in any of:
	/usr/local/go/src/github.com/algorand/go-algorand/test/partitiontest (from $GOROOT)
	/Users/jack/projects/algojack.go-algorand/cmd/partitiontest_linter/analyzer/testdata/src/github.com/algorand/go-algorand/test/partitiontest (from $GOPATH)
/Users/jack/projects/algojack.go-algorand/cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go:7:2: could not import github.com/algorand/go-algorand/test/partitiontest (invalid package name: "")
```

File: cmd/partitiontest_linter/analyzer/analyzer_test.go
Comment (algojack) [reply cmd/partitiontest_linter/analyzer/analyzer_test.go:1]: this one also fails for me:
```
Running tool: /usr/local/go/bin/go test -timeout 30s -coverprofile=/var/folders/m7/v0208hbj5p7643101fcjkpl00000gp/T/vscode-go78tJNa/go-code-cover github.com/algorand/go-algorand/cmd/partitiontest_linter/analyzer

# github.com/algorand/go-algorand/cmd/partitiontest_linter/analyzer [github.com/algorand/go-algorand/cmd/partitiontest_linter/analyzer.test]
/Users/jack/projects/algojack.go-algorand/cmd/partitiontest_linter/analyzer/analyzer_test.go:11:47: undefined: analyzer
FAIL	github.com/algorand/go-algorand/cmd/partitiontest_linter/analyzer [build failed]
FAIL
```

Comment (algorandskiy) [reply cmd/partitiontest_linter/analyzer/analyzer_test.go:1]: if `analyzer.go` and `analyzer_test.go` are in the same package, `Analyzer` var does not need package qualifier.

File: cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go
Comment (algojack) [reply cmd/partitiontest_linter/analyzer/testdata/analyzer_testdata_test.go:12]: updated with a comment, as this is input file.

File: cmd/partitiontest_linter/analyzer/analyzer_test.go
Comment (algojack) [reply cmd/partitiontest_linter/analyzer/analyzer_test.go]: fixed, thank you

**Comment by algojack** [comment]
not ready for merge yet, still need to hook into .golangci.yml

File: go.mod
Comment (algorandskiy) [inline go.mod:49]: I'm not sure about but this dep can go to buildtool installation script since it is not needed for building and running algod.

**Review by algorandskiy — COMMENTED** [review]
Looks much better, but still need to figure out what to do with the binary

File: go.mod
Comment (algojack) [reply go.mod]: agreed, removed.

**Comment by algojack** [comment]
> Looks much better, but still need to figure out what to do with the binary

so i built it on my machine for now. But depending on where we run it, we should be able to build it there as needed. We have a few options, but open to suggestions too:
1. don't build binary executable plugin.so file == build it where we use it
2. build it and include in this repo (not something we usually do)
3. build it and have a separate repo for plugin(s), so we just pull it from there to where we use it

**Comment by cce** [comment]
IMO I don't think we should include a binary .so file in the repo. We can just build it when we need it.

**Comment by algojack** [comment]
removed binary. To build it i did:
```
cd cmd/partitiontest_linter/
go build -buildmode=plugin -trimpath plugin/plugin.go
```

**Review by cce — DISMISSED** [review]
LGTM, after we merge this we can pull it into #2647 and figure out how to set up different golangci-lint configs for CI vs local, required vs suggested, etc.

**Review by algorandskiy — DISMISSED** [review]

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**
